mis/solver/greedyMIS.py

```diff
@@ -0,0 +1,134 @@
+                    combined_solution = MISSolution(original=graph, nodes=combined_nodes, energy=energy)
+
+                    if (best_solution is None) or (energy < best_solution.energy):
+                        best_solution = combined_solution
+
+        if best_solution is None:
+            return Execution.success([MISSolution(original=graph, nodes=[], energy=0)])
+        return Execution.success([best_solution])
+
+    def _exact_mis(self, instance: MISInstance) -> Execution[List[MISSolution]]:
+        graph = instance.graph
+        best: List[int] = []
+        self._backtrack(graph, [], 0, [best])
+        energy = -self._calculate_weight(best)
+        solution = MISSolution(original=graph, nodes=best, energy=energy)
+        return Execution.success([solution])
+
+    def _backtrack(self, graph: nx.Graph, subset: List[int], index: int, best: List[List[int]]) -> None:
+        nodes = list(graph.nodes())
+        if self._is_independent(graph, subset) and self._calculate_weight(subset) > self._calculate_weight(best[0]):
+            best[0] = subset[:]
+        for i in range(index, len(nodes)):
+            subset.append(nodes[i])
+            self._backtrack(graph, subset, i + 1, best)
+            subset.pop()
+
+    def _generate_subgraphs(self, graph: nx.Graph) -> List[Dict[int, int]]:
+        mappings = []
+        for node in graph.nodes():
+            mapper = GreedyMapping(
+                instance=MISInstance(graph),
+                lattice=self.lattice,
+                previously_generated_subgraphs=mappings,
+                seed=self.seed,
+            )
+            mapping = mapper.generate(starting_node=node)
+            mappings.append(mapping)
+        return sorted(mappings, key=lambda m: len(m), reverse=True)[:self.config.subgraph_quantity]
+
+    def _generate_lattice_graph(self, graph: nx.Graph, mapping: Dict[int, int]) -> nx.Graph:
+        G = nx.Graph()
+        for logical, physical in mapping.items():
+            weight = graph.nodes[logical].get("weight", 1.0)
+            pos = self.lattice.lattice.nodes[physical].get("pos", (0, 0))
+            G.add_node(physical, weight=weight, pos=pos)
+
+        for _, physical in mapping.items():
+            for neighbor in self.lattice.lattice.neighbors(physical):
+                if neighbor in mapping.values():
+                    G.add_edge(physical, neighbor)
+
+        return G
+
+    def _remove_nodes(self, graph: nx.Graph, nodes: List[int]) -> nx.Graph:
+        reduced = graph.copy()
+        to_remove = set(nodes)
+        for node in nodes:
+            to_remove.update(graph.neighbors(node))
+        reduced.remove_nodes_from(to_remove)
+        return reduced
+
+    def _calculate_weight(self, nodes: List[int]) -> float:
+        return sum(self.original_instance.graph.nodes[n].get("weight", 1.0) for n in nodes)
+
+    def _is_independent(self, graph: nx.Graph, nodes: List[int]) -> bool:
```

Comment:
Maybe this should become a method of `BaseSolver` ?

Reply:
I think that we have this method also in preprocessing (or perhaps postprocessing, I'm not sure). This should probably move to a module `graph_tools`?